invenio_rest/views.py

```diff
@@ -146,37 +150,66 @@ def get_method_serializers(self, http_method):
         )
 
     def match_serializers(self, serializers, default_media_type):
-        """Choose serializer for a given request based on accept headers.
+        """Choose serializer for a given request based on query arg or headers.
+
+        Checks if query arg `format` (by default) is present and tries to match
+        the serializer based on the arg value, by resolving the mimetype mapped
+        to the arg value.
+        Otherwise, chooses the serializer by retrieving the best quality
+        `Accept` headers and matching its value (mimetype).
 
         :param serializers: Dictionary of serializers.
         :param default_media_type: The default media type.
-        :returns: Best matching serializer based on client accept headers.
+        :returns: Best matching serializer based on `format` query arg first,
+            then client `Accept` headers or None if no matching serializer.
         """
-        # Bail out fast if no accept headers were given.
-        if len(request.accept_mimetypes) == 0:
-            return serializers[default_media_type]
-
-        # Determine best match based on quality.
-        best_quality = -1
-        best = None
-        has_wildcard = False
-        for client_accept, quality in request.accept_mimetypes:
-            if quality <= best_quality:
-                continue
-            if client_accept == '*/*':
-                has_wildcard = True
-            for s in serializers.keys():
-                if s in ['*/*', client_accept] and quality > 0:
-                    best_quality = quality
-                    best = s
-
-        # If no match found, but wildcard exists the use default media type.
-        if best is None and has_wildcard:
-            best = default_media_type
-
-        if best is not None:
-            return serializers[best]
-        return None
+        def _match_by_query_arg():
+            """Match serializer by query arg."""
+            def _serializer_exists_for_alias(alias):
```

Comment:
Why do we need to create a function if it is used only once and not visible from outside?
Calling functions in python is quite costly and this one will be called for every HTTP request.

Reply:
Because it is more human readable, and our code (I am sorry) it is not much!
Complex code should often be replaced by well defined names to immediately understand what's happening.
Have a look to this nested IFs for example:
https://github.com/inveniosoftware/invenio-rest/blob/master/invenio_rest/views.py#L110
This is not easily understandable I would immediately refactor it splitting it in several functions!

https://softwareengineering.stackexchange.com/questions/229610/indirection-readability-vs-performance

We can discuss this when you want!

Reply:
function removed because of comment below about EAFP.